### PR TITLE
Check unsupported TOP_LEVEL_KEYS

### PR DESCRIPTION
**Pull request by mitzkia**
This is a fix for: #965 

Of course I will also add unit-tests here.

Signed-off-by: Andras Mitzki <andras.mitzki@balabit.com>

### CODE REVIEW

**Comment by mitzkia** [comment]
Maybe the warning message not the best, not concrete.

File: pre_commit/repository.py [hunk 1]
```diff
@@ -190,6 +191,21 @@ def _cloned_repository_hooks(repo_config, store, root_config):
     )
 
 
+_TOP_LEVEL_KEYS = ('repo', 'rev', 'hooks')
```
Comment (asottile) [inline pre_commit/repository.py]: ideally this should be derived from `CONFIG_REPO_SCHEMA` in `pre_commit.clientlib` -- it probably also makes sense to implement this check as a `cfgv` checker in that module

also these aren't top level keys, just repository-level keys

File: pre_commit/repository.py [hunk 2]
```diff
@@ -190,6 +191,21 @@ def _cloned_repository_hooks(repo_config, store, root_config):
...
+
+
+def _check_extra_top_level_keys(repo_config):
+    extra_top_level_keys = set(dict(repo_config).keys()) - set(_TOP_LEVEL_KEYS)
```
Comment (asottile) [inline pre_commit/repository.py]: there's 4 unnecessary copies happening here:

1. `dict(repo_config)`
2. `.keys()` (in python 2)
3. `set(dict...)`
4. `set(_TOP_LEVEL_KEYS)`

The module constant should be `frozenset((...))`, you can use `set(repo_config)` directly

File: pre_commit/repository.py [hunk 3]
```diff
@@ -190,6 +191,21 @@ def _cloned_repository_hooks(repo_config, store, root_config):
...
+    if extra_top_level_keys:
+        logger.warning(
+            'Unexpected top level keys present: {}, only supported'
+            'top level keys: {}'.format(
```
Comment (asottile) [inline pre_commit/repository.py]: this produces a not-great error message that contains python literals, use `', '.join(sorted(...))` when displaying lists

**Comment by mitzkia** [comment]
I was also thinking about the exiting if there is an unexpected top level key. 
Could you help me to decide?

File: pre_commit/repository.py [hunk 4]
```diff
@@ -190,6 +191,21 @@ def _cloned_repository_hooks(repo_config, store, root_config):
...
 
 
+_TOP_LEVEL_KEYS = ('repo', 'rev', 'hooks')
+
+
+def _check_extra_top_level_keys(repo_config):
+    extra_top_level_keys = set(dict(repo_config).keys()) - set(_TOP_LEVEL_KEYS)
+    if extra_top_level_keys:
+        logger.warning(
+            'Unexpected top level keys present: {}, only supported'
```
Comment (asottile) [inline pre_commit/repository.py]: let's use consistent messaging with the warning on hook-dict extra keys (it's in this module under `create`)

**Comment by mitzkia** [comment]
Thanks, for the review notes. I will fix them. 
One more question: should not we exit if there is an unexpected key, or we can run forward?

**Comment by asottile** [comment]
> Thanks, for the review notes. I will fix them.
> One more question: should not we exit if there is an unexpected key, or we can run forward?

can't error without warning first, don't want to break 5 years of behaviour

**Comment by mitzkia** [comment]
Just found `cfgv.NoAdditionalKeys()`: which is a checker and proper message also displayed, maybe this is the function what we need here.

**Comment by asottile** [comment]
yeah you'd want something like that but `WarnAdditionalKeys` instead

**Comment by asottile** [comment]
@mitzkia with cfgv 1.6 you should be able to finish this :+1:

**Comment by asottile** [comment]
via #1028

**Pull request closed without merging**